Panic happends in kube-apiserver

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**: 
In kubernetes 1.15, Panic happends in kube-apiserver. Subsets[0] is  nil  when endpoints `kubernetes` has no subsets

The Panic happens in here:

https://github.com/kubernetes/kubernetes/blob/release-1.15/pkg/master/reconcilers/mastercount.go#L155


And everything returns to normal when we delete the endpoints `kubernetes` in  etcd 

**What you expected to happen**:
Maybe kube-apiserver should print error,not panic

**How to reproduce it (as minimally and precisely as possible)**:
The host's iowait of ETCD is high

**Anything else we need to know?**:
```
This is the log:
E1106 15:09:30.272637      12 prometheus.go:55] failed to register depth metric admission_quota_controller: duplicate metrics collector registration attempted
E1106 15:09:30.272705      12 prometheus.go:68] failed to register adds metric admission_quota_controller: duplicate metrics collector registration attempted
E1106 15:09:30.272773      12 prometheus.go:82] failed to register latency metric admission_quota_controller: duplicate metrics collector registration attempted
E1106 15:09:30.272814      12 prometheus.go:96] failed to register workDuration metric admission_quota_controller: duplicate metrics collector registration attempted
E1106 15:09:30.272854      12 prometheus.go:112] failed to register unfinished metric admission_quota_controller: duplicate metrics collector registration attempted
E1106 15:09:30.272888      12 prometheus.go:126] failed to register unfinished metric admission_quota_controller: duplicate metrics collector registration attempted
E1106 15:09:30.272938      12 prometheus.go:152] failed to register depth metric admission_quota_controller: duplicate metrics collector registration attempted
E1106 15:09:30.272968      12 prometheus.go:164] failed to register adds metric admission_quota_controller: duplicate metrics collector registration attempted
E1106 15:09:30.273033      12 prometheus.go:176] failed to register latency metric admission_quota_controller: duplicate metrics collector registration attempted
E1106 15:09:30.273160      12 prometheus.go:188] failed to register work_duration metric admission_quota_controller: duplicate metrics collector registration attempted
E1106 15:09:30.273214      12 prometheus.go:203] failed to register unfinished_work_seconds metric admission_quota_controller: duplicate metrics collector registration attempted
E1106 15:09:30.273248      12 prometheus.go:216] failed to register longest_running_processor_microseconds metric admission_quota_controller: duplicate metrics collector registration attempted
I1106 15:09:30.273284      12 plugins.go:158] Loaded 6 mutating admission controller(s) successfully in the following order: NamespaceLifecycle,LimitRanger,ServiceAccount,PodNodeSelector,DefaultTolerationSeconds,MutatingAdmissionWebhook.
I1106 15:09:30.273301      12 plugins.go:161] Loaded 6 validating admission controller(s) successfully in the following order: NamespaceExists,LimitRanger,ServiceAccount,PodNodeSelector,ValidatingAdmissionWebhook,ResourceQuota.
I1106 15:09:30.278022      12 client.go:354] parsed scheme: ""
I1106 15:09:30.278052      12 client.go:354] scheme "" not registered, fallback to default scheme
I1106 15:09:30.278124      12 asm_amd64.s:1337] ccResolverWrapper: sending new addresses to cc: [{192.17.102.86:4001 0  <nil>}]
I1106 15:09:30.278248      12 asm_amd64.s:1337] balancerWrapper: got update addr from Notify: [{192.17.102.86:4001 <nil>} {192.17.102.87:4001 <nil>} {192.17.102.88:4001 <nil>}]
I1106 15:09:30.344607      12 asm_amd64.s:1337] balancerWrapper: got update addr from Notify: [{192.17.102.88:4001 <nil>}]
W1106 15:09:30.344809      12 asm_amd64.s:1337] Failed to dial 192.17.102.86:4001: grpc: the connection is closing; please retry.
W1106 15:09:30.344883      12 asm_amd64.s:1337] Failed to dial 192.17.102.87:4001: grpc: the connection is closing; please retry.
I1106 15:09:30.346661      12 client.go:354] parsed scheme: ""
I1106 15:09:30.346688      12 client.go:354] scheme "" not registered, fallback to default scheme
I1106 15:09:30.346769      12 asm_amd64.s:1337] ccResolverWrapper: sending new addresses to cc: [{192.17.102.86:4001 0  <nil>}]
I1106 15:09:30.346883      12 asm_amd64.s:1337] balancerWrapper: got update addr from Notify: [{192.17.102.86:4001 <nil>} {192.17.102.87:4001 <nil>} {192.17.102.88:4001 <nil>}]
I1106 15:09:30.412213      12 asm_amd64.s:1337] balancerWrapper: got update addr from Notify: [{192.17.102.87:4001 <nil>}]
W1106 15:09:30.412395      12 asm_amd64.s:1337] Failed to dial 192.17.102.88:4001: grpc: the connection is closing; please retry.
W1106 15:09:30.412489      12 asm_amd64.s:1337] Failed to dial 192.17.102.86:4001: grpc: the connection is closing; please retry.
I1106 15:09:31.144084      12 trace.go:81] Trace[1019541866]: "List etcd3: key=/apiregistration.k8s.io/apiservices, resourceVersion=, limit: 10000, continue: " (started: 2019-11-06 15:09:30.345007277 +0800 CST m=+21.078109341) (total time: 799.016703ms):
Trace[1019541866]: [799.016703ms] [799.016703ms] END
I1106 15:09:33.283182      12 secure_serving.go:116] Serving securely on 192.17.102.89:4443
I1106 15:09:33.283295      12 autoregister_controller.go:140] Starting autoregister controller
I1106 15:09:33.283315      12 cache.go:32] Waiting for caches to sync for autoregister controller
I1106 15:09:33.284040      12 apiservice_controller.go:94] Starting APIServiceRegistrationController
I1106 15:09:33.284078      12 cache.go:32] Waiting for caches to sync for APIServiceRegistrationController controller
I1106 15:09:33.284107      12 controller.go:81] Starting OpenAPI AggregationController
I1106 15:09:33.284301      12 crdregistration_controller.go:112] Starting crd-autoregister controller
I1106 15:09:33.284332      12 controller_utils.go:1029] Waiting for caches to sync for crd-autoregister controller
I1106 15:09:33.285040      12 available_controller.go:376] Starting AvailableConditionController
I1106 15:09:33.285066      12 cache.go:32] Waiting for caches to sync for AvailableConditionController controller
I1106 15:09:33.286814      12 controller.go:83] Starting OpenAPI controller
I1106 15:09:33.288140      12 customresource_discovery_controller.go:208] Starting DiscoveryController
I1106 15:09:33.391393      12 crd_finalizer.go:255] Starting CRDFinalizer
I1106 15:09:33.391474      12 naming_controller.go:288] Starting NamingConditionController
I1106 15:09:33.391501      12 establishing_controller.go:73] Starting EstablishingController
I1106 15:09:33.391524      12 nonstructuralschema_controller.go:191] Starting NonStructuralSchemaConditionController
I1106 15:09:33.401853      12 log.go:172] http: TLS handshake error from 192.17.102.89:35940: EOF
I1106 15:09:33.408952      12 log.go:172] http: TLS handshake error from 192.17.102.89:35946: EOF
I1106 15:09:33.418395      12 log.go:172] http: TLS handshake error from 192.17.102.89:35996: EOF
I1106 15:09:33.443514      12 log.go:172] http: TLS handshake error from 192.17.102.89:35952: EOF
I1106 15:09:33.458631      12 log.go:172] http: TLS handshake error from 192.17.102.89:35970: EOF
I1106 15:09:33.461112      12 log.go:172] http: TLS handshake error from 192.17.102.89:35982: EOF
I1106 15:09:33.469309      12 log.go:172] http: TLS handshake error from 192.17.102.89:35960: EOF
I1106 15:09:33.477257      12 log.go:172] http: TLS handshake error from 192.17.102.89:35964: EOF
I1106 15:09:33.483561      12 cache.go:39] Caches are synced for autoregister controller
I1106 15:09:33.484269      12 cache.go:39] Caches are synced for APIServiceRegistrationController controller
I1106 15:09:33.484727      12 controller_utils.go:1036] Caches are synced for crd-autoregister controller
I1106 15:09:33.485215      12 cache.go:39] Caches are synced for AvailableConditionController controller
E1106 15:09:33.525582      12 runtime.go:73] Observed a panic: "index out of range" (runtime error: index out of range)
goroutine 2924 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x3c8d0a0, 0xa51dd10)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:69 +0x7b
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51 +0x82
panic(0x3c8d0a0, 0xa51dd10)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
k8s.io/kubernetes/pkg/master/reconcilers.(*masterCountEndpointReconciler).RemoveEndpoints(0xc000768e10, 0x4488460, 0xa, 0xc00074e930, 0x10, 0x10, 0xc00ba86000, 0x1, 0x1, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/master/reconcilers/mastercount.go:155 +0x5e0
k8s.io/kubernetes/pkg/master.(*Controller).Start(0xc000b20a80)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/master/controller.go:147 +0x1d8
k8s.io/kubernetes/pkg/master.(*Controller).PostStartHook(...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/master/controller.go:129
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.runPostStartHook.func1(0xc000b2e0b0, 0xc0016fb000, 0xb33, 0xc001662de0, 0xc0003705a0, 0xc0000dd7a0, 0xc003187f60)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/hooks.go:191 +0x6b
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.runPostStartHook(0x44a8583, 0x14, 0xc000b2e0b0, 0xc0016fb000, 0xb33, 0xc001662de0, 0xc0003705a0, 0xc0000dd7a0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/hooks.go:192 +0x85
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.(*GenericAPIServer).RunPostStartHooks
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/hooks.go:158 +0x13f
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 2924 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x3c8d0a0, 0xa51dd10)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
k8s.io/kubernetes/pkg/master/reconcilers.(*masterCountEndpointReconciler).RemoveEndpoints(0xc000768e10, 0x4488460, 0xa, 0xc00074e930, 0x10, 0x10, 0xc00ba86000, 0x1, 0x1, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/master/reconcilers/mastercount.go:155 +0x5e0
k8s.io/kubernetes/pkg/master.(*Controller).Start(0xc000b20a80)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/master/controller.go:147 +0x1d8
k8s.io/kubernetes/pkg/master.(*Controller).PostStartHook(...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/master/controller.go:129
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.runPostStartHook.func1(0xc000b2e0b0, 0xc0016fb000, 0xb33, 0xc001662de0, 0xc0003705a0, 0xc0000dd7a0, 0xc003187f60)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/hooks.go:191 +0x6b
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.runPostStartHook(0x44a8583, 0x14, 0xc000b2e0b0, 0xc0016fb000, 0xb33, 0xc001662de0, 0xc0003705a0, 0xc0000dd7a0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/hooks.go:192 +0x85
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.(*GenericAPIServer).RunPostStartHooks
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/hooks.go:158 +0x13f
W1106 15:11:19.139832      12 feature_gate.go:208] Setting GA feature gate HugePages=true. It will be removed in a future release.
I1106 15:11:19.139933      12 feature_gate.go:216] feature gates: &{map[CustomResourceSubresources:true HugePages:true]}
Flag --insecure-port has been deprecated, This flag will be removed in a future version.
Flag --admission-control has been deprecated, Use --enable-admission-plugins or --disable-admission-plugins instead. Will be removed in a future version.
Flag --experimental-encryption-provider-config has been deprecated, use --encryption-provider-config.
I1106 15:11:19.140089      12 flags.go:33] FLAG: --address="127.0.0.1"
I1106 15:11:19.140119      12 flags.go:33] FLAG: --admission-control="[ValidatingAdmissionWebhook,MutatingAdmissionWebhook,NamespaceLifecycle,NamespaceExists,LimitRanger,ResourceQuota,ServiceAccount,PodNodeSelector,DefaultTolerationSeconds]"
I1106 15:11:19.140147      12 flags.go:33] FLAG: --admission-control-config-file=""
I1106 15:11:19.140155      12 flags.go:33] FLAG: --advertise-address="192.17.102.91"
I1106 15:11:19.140161      12 flags.go:33] FLAG: --allow-privileged="true"
I1106 15:11:19.140168      12 flags.go:33] FLAG: --alsologtostderr="false"
I1106 15:11:19.140175      12 flags.go:33] FLAG: --anonymous-auth="false"
I1106 15:11:19.140181      12 flags.go:33] FLAG: --api-audiences="[]"


```
**Environment**:
- Kubernetes version (use `kubectl version`):

```bash
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.0", GitCommit:"e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529", GitTreeState:"clean", BuildDate:"2019-06-19T16:40:16Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}

Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.0", GitCommit:"e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529", GitTreeState:"clean", BuildDate:"2019-06-19T16:40:16Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:NA
- OS (e.g: `cat /etc/os-release`):
```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
- Kernel (e.g. `uname -a`):
```
3.10.0-693.el7.x86_64
```
- Install tools: `kubeadm`
- Network plugin and version (if this is a network-related bug):
- Others:

--- a/pkg/master/controller_test.go
+++ b/pkg/master/controller_test.go
@@ -558,6 +558,32 @@ func TestReconcileEndpoints(t *testing.T) {
 
 }
 
+func TestEmptySubsets(t *testing.T) {
+	ns := metav1.NamespaceDefault
+	om := func(name string) metav1.ObjectMeta {
+		return metav1.ObjectMeta{Namespace: ns, Name: name}
+	}
+	endpoints := &corev1.EndpointsList{
+		Items: []corev1.Endpoints{{
+			ObjectMeta: om("foo"),
+			Subsets:    nil,
+		}},
+	}
+	fakeClient := fake.NewSimpleClientset()
+	if endpoints != nil {
+		fakeClient = fake.NewSimpleClientset(endpoints)
+	}
+	epAdapter := reconcilers.NewEndpointsAdapter(fakeClient.CoreV1(), nil)
+	reconciler := reconcilers.NewMasterCountEndpointReconciler(1, epAdapter)
+	endpointPorts := []corev1.EndpointPort{
+		{Name: "foo", Port: 8080, Protocol: "TCP"},
+	}
+	err := reconciler.RemoveEndpoints("foo", net.ParseIP("1.2.3.4"), endpointPorts)
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+}
+
 func TestCreateOrUpdateMasterService(t *testing.T) {
 	ns := metav1.NamespaceDefault
 	om := func(name string) metav1.ObjectMeta {

--- a/pkg/master/reconcilers/lease.go
+++ b/pkg/master/reconcilers/lease.go
@@ -67,9 +67,11 @@ func (s *storageLeases) ListLeases() ([]string, error) {
 		return nil, err
 	}
 
-	ipList := make([]string, len(ipInfoList.Items))
-	for i, ip := range ipInfoList.Items {
-		ipList[i] = ip.Subsets[0].Addresses[0].IP
+	ipList := make([]string, 0, len(ipInfoList.Items))
+	for _, ip := range ipInfoList.Items {
+		if len(ip.Subsets) > 0 && len(ip.Subsets[0].Addresses) > 0 && len(ip.Subsets[0].Addresses[0].IP) > 0 {
+			ipList = append(ipList, ip.Subsets[0].Addresses[0].IP)
+		}
 	}
 
 	klog.V(6).Infof("Current master IPs listed in storage are %v", ipList)

--- a/pkg/master/reconcilers/lease_test.go
+++ b/pkg/master/reconcilers/lease_test.go
@@ -617,6 +617,19 @@ func TestLeaseRemoveEndpoints(t *testing.T) {
 				}},
 			},
 		},
+		{
+			testName:      "endpoint with no subset",
+			serviceName:   "foo",
+			ip:            "5.6.7.8",
+			endpointPorts: []corev1.EndpointPort{{Name: "foo", Port: 8080, Protocol: "TCP"}},
+			endpointKeys:  []string{"1.2.3.4", "4.3.2.2", "4.3.2.3", "4.3.2.4"},
+			endpoints: &corev1.EndpointsList{
+				Items: []corev1.Endpoints{{
+					ObjectMeta: om("foo"),
+					Subsets:    nil,
+				}},
+			},
+		},
 	}
 	for _, test := range stopTests {
 		t.Run(test.testName, func(t *testing.T) {

--- a/pkg/master/reconcilers/mastercount.go
+++ b/pkg/master/reconcilers/mastercount.go
@@ -149,6 +149,10 @@ func (r *masterCountEndpointReconciler) RemoveEndpoints(serviceName string, ip n
 		return err
 	}
 
+	if len(e.Subsets) == 0 {
+		// no action is needed to remove the endpoint
+		return nil
+	}
 	// Remove our IP from the list of addresses
 	new := []corev1.EndpointAddress{}
 	for _, addr := range e.Subsets[0].Addresses {

